File: Sources/YTools/Modules/Dictionary/SpellingModule.swift
import AppKit
import YToolsModuleKit

struct SpellingModule: YToolsModule {
    let descriptor = ModuleDescriptor(id: "spelling", name: "拼写检查")

    func search(_ request: ModuleSearchRequest) async throws -> [LauncherResult] {
        let trimmed = request.query.trimmingCharacters(in: .whitespacesAndNewlines)
        let prefixes = ["spell", "拼写"]
        guard let prefix = prefixes.first(where: {
            trimmed.lowercased().hasPrefix($0.lowercased() + " ")
        }) else { return [] }
        let word = String(trimmed.dropFirst(prefix.count)).trimmingCharacters(in: .whitespacesAndNewlines)
        guard !word.isEmpty else { return [] }

        let outcome = await MainActor.run { () -> (guesses: [String], isCorrect: Bool) in
            let checker = NSSpellChecker.shared
            let range = NSRange(location: 0, length: (word as NSString).length)
            let guesses = checker.guesses(
                forWordRange: range,
                in: word,
                language: nil,
                inSpellDocumentWithTag: 0
            ) ?? []
            return (guesses, checker.checkSpelling(of: word, startingAt: 0).location == NSNotFound)
        }
        if outcome.guesses.isEmpty, outcome.isCorrect {
            return [LauncherResult(
                id: "spelling:correct:\(word)",
                moduleID: descriptor.id,
                title: word,
                subtitle: "拼写正确 · 回车复制",
                icon: .system("checkmark.seal"),
                score: 1_050,
                action: .copy(word)
            )]
        }
        return outcome.guesses.prefix(8).enumerated().map { index, suggestion in
            LauncherResult(
                id: "spelling:\(word):\(suggestion)",
                moduleID: descriptor.id,
                title: suggestion,
                subtitle: "“\(word)” 的拼写建议 · 回车复制",
                icon: .system("textformat.abc.dottedunderline"),
                score: 1_050 - index,
                action: .copy(suggestion)
            )
        }
    }
}

File: src/YTools.Windows/Services/Modules/SpellingModule.cs
using YTools.ModuleKit;

namespace YTools.Services.Modules;

public sealed class SpellingModule : IYToolsModule
{
    private readonly SpellingService _spelling;

    public SpellingModule(SpellingService spelling)
    {
        _spelling = spelling;
    }

    public ModuleDescriptor Descriptor { get; } = new("spelling", "拼写检查");

    public Task<IReadOnlyList<LauncherResult>> SearchAsync(ModuleSearchRequest request)
    {
        var trimmed = request.Query.Trim();
        var prefixes = new[] { "spell", "拼写" };
        var prefix = prefixes.FirstOrDefault(candidate =>
            trimmed.StartsWith(candidate + " ", StringComparison.OrdinalIgnoreCase));
        if (prefix is null)
        {
            return Task.FromResult<IReadOnlyList<LauncherResult>>([]);
        }

        var word = trimmed[prefix.Length..].Trim();
        if (string.IsNullOrEmpty(word))
        {
            return Task.FromResult<IReadOnlyList<LauncherResult>>([]);
        }

        var outcome = _spelling.Check(word);
        if (outcome.IsCorrect)
        {
            return Task.FromResult<IReadOnlyList<LauncherResult>>(
            [
                new LauncherResult(
                    $"spelling:correct:{word}",
                    Descriptor.Id,
                    word,
                    "拼写正确 · 回车复制",
                    new ResultIcon.System("checkmark.seal"),
                    1_050,
                    new ResultAction.Copy(word))
            ]);
        }

        var results = outcome.Suggestions.Select((suggestion, index) =>
            new LauncherResult(
                $"spelling:{word}:{suggestion}",
                Descriptor.Id,
                suggestion,
                $"“{word}” 的拼写建议 · 回车复制",
                new ResultIcon.System("textformat.abc.dottedunderline"),
                1_050 - index,
                new ResultAction.Copy(suggestion))).Cast<LauncherResult>().ToList();
        return Task.FromResult<IReadOnlyList<LauncherResult>>(results);
    }
}
